### ISSUE TITLE
Better way to integrate filters

### ISSUE THREAD

**Issue by aristotelesneto** [logsearch/logsearch-boshrelease#9]
Hi David,

Would you be interested if I went away and formulated a better way to integrate the filters into the main logsearch-boshrelease, so we don't need to have a separate filter bosh release?

I was thinking about something along the lines of leveraging the pre_packaging / packaging scripts, along with a job that (if configured to do so), would periodically check that the filters are up to date with a remote git repo (from the deployed machine).

Thoughts?

**Comment by mrdavidlaing** [comment]
YES!

What we currently have is kinda crummy; so any improvements to the flow much appreciated.

The design goals are:
1.  Deploying logsearch-boshrelease gets you set the set of https://github.com/logsearch/logsearch-filters-common filters
2.  You should then be able to "extend" your logsearch deployment by choosing to add some community filters - eg, https://github.com/logsearch/logstash-filters-cf
3.  To add your own custom filters, you should fork https://github.com/logsearch/logsearch-filters-common, replace the common with your own filters and then "extend" your logsearch deployment with those as-well.

In our experience, once you have your logsearch cluster running, updating filters is by _far_ the most common change.  Ideally we want an easily "continuous deployable" way to (a) edit `logsearch-filters-mine`, (b) run unit tests, and then (c) "bosh deploy" them into your existing cluster.

**Comment by mrdavidlaing** [comment]
@wdneto - I've given you commit access to the relevant repo's - please just make your PR branch in the repo itself.  Makes it easier for us to collaborate on PRs.

Please _don't_ merge to master

**Comment by aristotelesneto** [comment]
Cool, thanks. Will make sure the branch is in the logsearch repo.

Upon reading the design goals, the first thing that came to mind was the way buildpacks are now installed into cf as of v166, except without the whole api side of things. 

I'll have a play and see what I can come up with. 

Did you have any thoughts on a self-updating job or would you rather new rules still be pushed via bosh deploy (rather than pulled by the already deployed machine)?

**Comment by mrdavidlaing** [comment]
> Did you have any thoughts on a self-updating job or would you rather new rules still be pushed via bosh deploy (rather than pulled by the already deployed machine)?

@dpb587 and I are against a self updating job; since it makes auditing the current state of the system difficult.  I think a solution that requires a `bosh deploy` to be run is the right way to go.

**Comment by mrdavidlaing** [comment]
@wdneto Do you have any code / designs to share for this?

**Comment by dpb587** [comment]
I'm going to hack on ideas for this a bit today. I'll update or send a PR within a few hours.

**Comment by dpb587** [comment]
My idea was a dud. Thinking a bit more before trying anything further...

**Comment by aristotelesneto** [comment]
I'll try and get a doc up with some ideas. I'm taking the time on Easter holidays to pack up (I'm moving house soon), so that's in the way a bit. 

The main idea was to merge filters as sub modules inside the logsearch boshrelease, and leveraging prepackaging an packaging scripts to run the compilation / tests automatically during deploy.

**Comment by mrdavidlaing** [comment]
> The main idea was to merge filters as sub modules inside the logsearch-boshrelease

Would this not require each logsearch-boshrelease "deployment" to have its own fork of the  logsearch-boshrelease; and then create your own releases from your fork?

Would you not end up with release number conflicts when you pulling new upstream logsearch-boshrelease s into your fork?

**Comment by aristotelesneto** [comment]
It would only require a fork if you wanted to include additional filter repositories as submodules, but it shouldn't create numbering conflicts when picking up upstream changes (upstream being the main logsearch-boshrelease+filters). I can see that being undesirable (it was painful to maintain cf-release/dea when bundling additional system buildpacks in cf-release).

I think this can be overcome by allowing the job definition to pull "additional filters" at deploy-time, directly from a git url. This is sort of similar to the current way buildpacks are bundled in cf-release - there are system buildpacks (which are submodules [[1]](https://github.com/cloudfoundry/cf-release/tree/master/src/buildpacks) and [installed on-deploy](https://github.com/cloudfoundry/cf-release/tree/master/packages/buildpack_nodejs)), and there are admin buildpacks which are uploaded (cf create-buildpacks) - but instead of having upload functionality, the 'filter' job would simply pull directly off git (on a particular branch/commit sha), which would be defined on the deployment manifest.This is actually how I was intending on doing the self-updating filters, but that can be controlled via an optional property set on manifest (defaulting to false).

Thoughts?

**Comment by mrdavidlaing** [comment]
@wdneto - I can almost guarantee that you are going to want your own custom filters.

@dpb587, @wdneto - you're both converging on the 

> allowing the job definition to pull "additional filters" at deploy-time

technique.  

I still feel uncomfortable about introducing an external deploy time dependancy (ie, files not coming from the blobstore); it doesn't feel like the "BOSH way".

But implementation wise, its much simpler than separate filter boshreleases.  And that is a big :+1: in its favour.

As a final datapoint, lets think through the experience of consuming extra filters.  I'll start with the existing "filter boshrelease" process, could one of you do the same for the "the job definition to pull "additional filters" approach.

The use cases: 
1. I want to install a logsearch cluster with extra (a) cloudfoundry filters 
2. I want to upgrade the cloudfoundry filters
3. I want to add my own filters
4. I want to upgrade my own filters
5. I want to know what version of the filters I have installed

**Comment by mrdavidlaing** [comment]
## "filter boshrelease" process

#### 1. I want to install a logsearch cluster with extra (a) cloudfoundry filters
1. bosh upload release for logsearch-boshrelease
2. bosh upload release for logsearch-filters-cloudfoundry-boshrelease
3. create deploy manifest with 2 releases, and add extra job templates to existing log_parser 

```
releases:
- name: logsearch
  version: latest
- name: logsearch-cf_parsers
  version: latest
...snip...
- name: log_parser
  release: logsearch
  templates:
  - name: collectd
  - name: log_parser
  - name: logstash-filters-cf
    release: logsearch-cf_parsers
    ....
```
1. bosh deploy

#### 2. I want to upgrade the cloudfoundry filters
1. bosh upload release for logsearch-filters-cloudfoundry
2. bosh deploy

#### 3. I want to add my own filters
1. fork logsearch-filters-cloudfoundry -> logsearch-filters-NAME
2. fork logsearch-filters-cloudfoundry-boshrelease -> logsearch-filters-NAME-boshrelease
3. edit and compile logsearch-filters-NAME
4. add logsearch-filters-NAME output as logsearch-filters-NAME-boshrelease blob.  Update spec. bosh create release
5. bosh upload release
6. amend deploy manifest, (now has 3 releases, and 2 extra job templates to for each log_parser )
7. bosh deploy

#### 4. I want to upgrade my own filters
1. edit and compile logsearch-filters-NAME
2. add logsearch-filters-NAME output as logsearch-filters-NAME-boshrelease blob.  Update spec. bosh create release
3. bosh upload release
4. bosh deploy

#### 5. I want to know what version of the filters I have installed

```
$ bosh deployments

+------------------------+------------------------------+
| Name                   | Release(s)                   | 
+------------------------+------------------------------
| meta-logsearch-io      | logsearch/8                  |
|                        | logstash-filters-cf/1        |
+------------------------+------------------------------+
```

**Comment by mrdavidlaing** [comment]
## "allowing the job definition to pull additional filters at deploy-time" process

#### 1. I want to install a logsearch cluster with extra (a) cloudfoundry filters
1. bosh upload release for logsearch-boshrelease
2. Add download locations for parsers to `properties:`

```
properties:
  log_parser:
    filters:
    - http://ci-logsearch.s3-website-eu-west-1.amazonaws.com/logsearch-filters-cf.v1.tar.gz
```
1. bosh deploy

#### 2. I want to upgrade the cloudfoundry filters
1. edit deploy manifest, update download locations

```
properties:
  log_parser:
    filters:
    - http://ci-logsearch.s3-website-eu-west-1.amazonaws.com/logsearch-filters-cf.v2.tar.gz
```
1. bosh deploy

#### 3. I want to add my own filters
1. fork logsearch-filters-cloudfoundry 
2. edit and compile logsearch-filters-NAME
3. publish filters to HTTP endpoint (S3, Github releases)
4. edit deploy manifest, add extra download locations

```
properties:
  log_parser:
    filters:
    - http://ci-logsearch.s3-website-eu-west-1.amazonaws.com/logsearch-filters-cf.v2.tar.gz
    - https://github.com/myrepo/logsearch-filters-NAME/releases/download/v1.0/logsearch-filters-NAME.tar.gz
```
1. bosh deploy

#### 4. I want to upgrade my own filters
1. edit and compile logsearch-filters-NAME
2. publish filters to HTTP endpoint (S3, Github releases)
3. edit deploy manifest, add update download locations version

#### 5. I want to know what version of the filters I have installed
1. `cat deploy_manifest.yaml | grep -b5 filters`

**Comment by mrdavidlaing** [comment]
And, having written it all down, I'm now convinced that the "allowing the job definition to pull additional filters at deploy-time" process is both easier to implement and to consume.

**Comment by mrdavidlaing** [comment]
@dpb587, you care to have a go implementing the "allowing the job definition to pull additional filters at deploy-time" process ?

**Comment by dpb587** [comment]
I can have a go at it this afternoon or tomorrow.

**Comment by aristotelesneto** [comment]
> I still feel uncomfortable about introducing an external deploy time dependancy (ie, files not coming from the blobstore); it doesn't feel like the "BOSH way".

As far as I can tell, theres no way to accomplish this without having a requirement for people to maintain their own forks of the boshrelease, or continuously having to bump the release version with filter tweaks (which I'd expect to be a very frequent occurence).

> And, having written it all down, I'm now convinced that the "allowing the job definition to pull additional filters at deploy-time" process is both easier to implement and to consume.

I entirely agree with you that having external dependencies at deploy time, is far from an ideal way of maintaining a distributed system, which is why I was referencing the way cf-release is currently handling buildpacks. 

For the standard/default/system buildpacks, they are submodules and packages used by the ccng job. For user-provided buildpacks, they have to be uploaded (via cf create-buildpack), or a cf-release fork maintained and new submodules / packages added.

What I was trying to get at was that specifying filter URIs on the deployment manifest is simply substituting the 'cf create-buildpack' option, as a way to pull in arbitrary filters. If someone is deploying logsearch needs to remove the external dependency, then they would have to create and maintain a fork of the main boshrelease, and create their own package therein.

This way we have a mixture of both methods - a safe approach for filters we always want to have present (that come from the blobstore), and a flexible approach for custom filters that can be injected into the config from arbitrary locations.

**Comment by mrdavidlaing** [comment]
@wdneto - I'm a bit confused.  Do you have a third solution in mind, or are you talking about the "allowing the job definition to pull additional filters at deploy-time" process - P2 described above?

> For the standard/default/system buildpacks, they are submodules and packages used by the ccng job. 

In P2, [standard filters are a package in the logsearch-boshrelease](https://github.com/logsearch/logsearch-boshrelease/tree/master/packages/logstash-filters-common)

> What I was trying to get at was that specifying filter URIs on the deployment manifest is simply substituting the 'cf create-buildpack' option, as a way to pull in arbitrary filters.

Agreed.

> If someone is deploying logsearch needs to remove the external dependency, then they would have to create and maintain a fork of the main boshrelease, and create their own package therein.

Disagree.  In P2:
- Standard filters are part of the logsearch-boshrelease.  No external dependancy.
- You'd need to manually copy the extra filters from their external urls to some internal webserver; and change your deploy_manifest.yml to reference them on the internal webserver.  No changes to logsearch-boshrelease required.

> This way we have a mixture of both methods 

Which way?  P2 ?

**Comment by aristotelesneto** [comment]
I think we're confusing each other because of a discrepancy in the terminology we're using.

When I say "external dependency", I mean external to the distributed system being managed via BOSH, not necessarily external to the company. But with your definition it does make sense that the only thing needed would be the shift in urls to an internal webserver.

When you say "allowing the job definition to pull", the way I understand that as the actual job (that gets started via a ctl script) on a running machine, as opposed to the machine you're running bosh commands doing the pulling.

> In P2, standard filters are a package in the logsearch-boshrelease

But it's a blob stored in S3, so even though its a package, you're relying on s3 being available. A submodule would allow you to pull the code in, and the tarball built at 'bosh deploy' time.

**Comment by aristotelesneto** [comment]
Do you use IM or something, might be faster to hash this out? :P

**Comment by mrdavidlaing** [comment]
https://www.flowdock.com/app/cityindexlabs/logsearch-community

Make a flowdock account, and I'll invite you to the flow

**Comment by mrdavidlaing** [comment]
Send the email you registered with flowdock to david.laing@labs.cityindex.com

**Comment by mrdavidlaing** [comment]
> >  In P2, standard filters are a package in the logsearch-boshrelease
> 
> But it's a blob stored in S3, so even though its a package, you're relying on s3 being available. 

At bosh upload release time you need access to S3 to get all published logsearch-boshrelease packages; (which includes the logsearch-filters-common blobs, the elasticsearch blobs etc)

> When you say "allowing the job definition to pull", the way I understand that as the actual job (that gets started via a ctl script) on a running machine

Yep.  Can't see a way around that.  We should cache the download, so it only happens once.

**Comment by aristotelesneto** [comment]
I'll try setting something up, and see what it looks like.

**Comment by mrdavidlaing** [comment]
Released as part of #25 

Additional filters are now pulled from a Url specified in your deployment manifest:

```
   logstash_parser:
      filters:
     - "https://ci-logsearch.s3.amazonaws.com/logsearch-filters-cf/logstash-filters-cf-7.tgz"
```
